### PR TITLE
Add some more information to the inference data of tfp

### PR DESCRIPTION
**Pull request by yaochitc**
These parts are now supported:
1. observed data
2. posterior predictive
3. log-likelihood of observed data

More parameters should be passed to the from_tfp method, including:
1. model_fn: a model function
2. posterior_predictive_samples: how many samples to use from the chain
3. observed: observed values

### CODE REVIEW

**Comment by yaochitc** [comment]
ok

File: arviz/tests/helpers.py [hunk 1]
```diff
@@ -178,10 +178,18 @@ def target_log_prob_fn(avg_effect, avg_stddev, school_effects_standard):
         ),
     )
 
+    def model_fn():
```
[the remark this answers is not in the document]
Comment (yaochitc) [reply arviz/tests/helpers.py]: the model_fn is used to encapsulate the parameters needed to execute the model, which makes it easier to use in the arviz module.

File: arviz/data/io_tfp.py [hunk 2]
```diff
@@ -32,18 +52,128 @@ def posterior_to_xarray(self):
...
+            with ed.interception(self._value_setter(variables)):
+                if self.posterior_predictive_size > 1:
+                    posterior_preds.append(
+                        [self.model_fn() for _ in range(self.posterior_predictive_size)]
+                    )
+                else:
+                    posterior_preds.append(self.model_fn())
+
+        data = {}
+        with tf.Session() as sess:
```
[the remark this answers is not in the document]
Comment (yaochitc) [reply arviz/data/io_tfp.py]: > Out of curiosity: In `observed_data_to_xarray` lines 58 to 62 check the object type, and optionally run tensorflow compute if the values are tensors (I think)
> 
> Is that not needed in `posterior_predictive_to_xarray`?
> 
> https://github.com/arviz-devs/arviz/pull/447/files#diff-b618a7e0bc3e4dd03124ca9827bee67fR58

As the sess.run() is used to execute a graph, so an exception will be thrown if the parameter passed in is not a tensor.

File: arviz/tests/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/tests/helpers.py]: Ah got it, could this be changed to a lambda function?

File: arviz/data/io_tfp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/data/io_tfp.py]: Thank you for the explanation!

File: arviz/tests/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (canyon289) [reply arviz/tests/helpers.py]: @yaochitc Did you want to change to a lambda on this PR or should we merge and do that on another one?

**Comment by ahartikainen** [comment]
Could you also fix multichain situation?

    `np.expand_dims` vs `np.atleast_3d`

File: arviz/tests/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yaochitc) [reply arviz/tests/helpers.py]: > @yaochitc Did you want to change to a lambda on this PR or should we merge and do that on another one?

Ok, I'll do it in this pr.

**Comment by yaochitc** [comment]
> Could you also fix multichain situation?
> 
> ```
> `np.expand_dims` vs `np.atleast_3d`
> ```

Don't quite understand your meaning, could you give some explanation?

**Comment by ahartikainen** [comment]
Oh yes `np.atleast` is not going to work, so one needs to check the chain count and work based on that.

**Comment by yaochitc** [comment]
> so one needs t

I'll have a look a it.

**Comment by yaochitc** [comment]
@canyon289 Hello, the model_fn is changed to a lambda in this pr.

**Comment by ahartikainen** [comment]
@yaochitc good to know.

So to run 4 chains one would need to run that code 4 times or what does the `parallel_iterations` do?
Also if one would need to run each chain separately, then I think `from_tfp` could accept a list of chains?

**Comment by yaochitc** [comment]
> @yaochitc good to know.
> 
> So to run 4 chains one would need to run that code 4 times or what does the `parallel_iterations` do?
> Also if one would need to run each chain separately, then I think `from_tfp` could accept a list of chains?

The `parallel_iterations` is used to control the parallelism of loop which has no effect on the sampling result, and I think it is good to accept a list a chains for the method.

File: arviz/tests/helpers.py [hunk 3]
```diff
@@ -189,7 +189,12 @@ def target_log_prob_fn(avg_effect, avg_stddev, school_effects_standard):
     with tf.Session() as sess:
         [states_, _] = sess.run([states, kernel_results])
 
-    data = from_tfp(states_, var_names=["mu", "tau", "eta"])
+    data = from_tfp(
+        states_,
+        var_names=["mu", "tau", "eta"],
+        model_fn=lambda: schools_model(data["J"], data["sigma"].astype(np.float32)),
```
Comment (canyon289) [inline arviz/tests/helpers.py]: Thanks @yaochitc

**Comment by jeffpollock9** [comment]
Hi, I noticed that in this merge request the following lines were added to the top of [io_tfp.py](https://github.com/arviz-devs/arviz/pull/447/files#diff-b618a7e0bc3e4dd03124ca9827bee67fR4):

```python
import tensorflow as tf
import tensorflow_probability.python.edward2 as ed
```

Which I think means that tensorflow is now a requirement of arviz. Was this intended? I don't think people usually have pytorch and tensorflow in the same virtual environment, or at least I don't.

**Comment by ahartikainen** [comment]
This was not intented. I have not yet tested latest master, but I think this will fail without  tensorflow installed. (We need to test this too)

I think moving imports inside class like other `from_xyz` functions do, should work.

**Pull request merged**